pkg/metrics/garden.go

```diff
@@ -0,0 +1,288 @@
+		metric, err := prometheus.NewConstMetric(c.descs[metricGardenShootInfo], prometheus.GaugeValue, 0, shoot.Name, shoot.Namespace, iaas, shoot.Spec.Kubernetes.Version, shoot.Spec.Cloud.Region, seed)
+		if err != nil {
+			ScrapeFailures.With(prometheus.Labels{"kind": "shoots"}).Inc()
+			continue
+		}
+		ch <- metric
+
+		// Collect metrics to the node count of the Shoot.
+		// TODO: Use the metrics of the Machine-Controller-Manager, when available. The mcm should be able to provide the actual amount of nodes/machines.
+		c.collectShootNodeMetrics(*shoot, cloudProvider, ch)
+
+		if shoot.Status.LastOperation != nil {
+			lastOperation := string(shoot.Status.LastOperation.Type)
+			lastOperationState := string(shoot.Status.LastOperation.State)
+
+			// Export a metric for any possible operation, which can be ongoing on the Shoot.
+			// For currently non ongoing operations the value of the metric not will be set to 0.
+			for _, operation := range shootOperations {
+				var operationState float64
+				if operation == lastOperation {
+					switch shoot.Status.LastOperation.State {
+					case gardenv1beta1.ShootLastOperationStateSucceeded:
+						operationState = 1
+					case gardenv1beta1.ShootLastOperationStateProcessing:
+						operationState = 2
+					case gardenv1beta1.ShootLastOperationStatePending:
+						operationState = 3
+					case gardenv1beta1.ShootLastOperationStateError:
+						operationState = 4
+					case gardenv1beta1.ShootLastOperationStateFailed:
+						operationState = 5
+					}
+				}
+				metric, err := prometheus.NewConstMetric(c.descs[metricGardenShootOperationState], prometheus.GaugeValue, operationState, shoot.Name, shoot.Namespace, operation, mailTo)
+				if err != nil {
+					ScrapeFailures.With(prometheus.Labels{"kind": "shoots"}).Inc()
+					continue
+				}
+				ch <- metric
+			}
+
+			// Export a metric for each condition of the Shoot.
+			for _, condition := range shoot.Status.Conditions {
+				metric, err := prometheus.NewConstMetric(c.descs[metricGardenShootCondition], prometheus.GaugeValue, mapConditionStatus(condition.Status), shoot.Name, shoot.Namespace, string(condition.Type), lastOperation, mailTo)
+				if err != nil {
+					ScrapeFailures.With(prometheus.Labels{"kind": "shoots"}).Inc()
+					continue
+				}
+				ch <- metric
+			}
+
+			// Collect the current count of ongoing operations.
+			if !isSeed {
+				shootOperationsCounters[fmt.Sprintf("%s:%s:%s:%s:%s:%s", lastOperation, lastOperationState, iaas, seed, shoot.Spec.Kubernetes.Version, shoot.Spec.Cloud.Region)]++
+			}
+		}
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go shootValuesListToMetrics(shootCounters, c.descs[metricGardenShootsTotal], ch, &wg)
+	go shootValuesListToMetrics(shootOperationsCounters, c.descs[metricGardenOperationsTotal], ch, &wg)
+	wg.Wait()
+}
+
+func (c gardenMetricsCollector) collectShootNodeMetrics(shoot gardenv1beta1.Shoot, cloudProvider gardenv1beta1.CloudProvider, ch chan<- prometheus.Metric) {
```

Comment:
Use pointers for `gardenv1beta1.Shoot` and `gardenv1beta1.CloudProvider`

Reply:
Pointer for ``gardenv1beta1.Shoot`` makes sense. For ``cloudProvider`` not, because the ``collectShootNodeMetrics`` function will create then a ``github.com/gardener/gardener/pkg/operation/shoot.Shoot`` struct and it doesn't expect a pointer for ``CloudProvider`` here.